### PR TITLE
Feat: Party Skill Statistics

### PR DESCRIPTION
**Pull request by franneck94**
Hey,
I've added a new Window for the statistics of the skill usage of the team members, as suggested in #535.

The new feature consist of the following:

- Get the names of the players and their skills:
  - Load the skillbar for the player and his heroes
  - For other players or henchmen the skillbar cannot be loaded, therefore the skills are initialized with "Unknown Skill" 
   - When entering an explorable and the other player/henchman will activate their skills, these will then replace the beforehand "Unknown Skills"
- After leaving an explorable and going back to the outpost, the data from the previous run will be displayed until any party member leaves or a new party member joins

Example:

Screenshot_8

For every player/hero/henchman in the party:
- The absolute number of activations for each skill, if the setting for displaying the absolute number is active
- The percentage of activations for each skill, if the setting for displaying the percentage value is active

These stats can be posted in the chat by the following commands:
- /skillstats: Will display the stats for the player
- /skillstats 1: Will display the stats for the team member at position 1
- /skillstats 1 1: Will display the stats for the first skill of the first team member
- /skillstats reset: Will reset the statistics

Note: Skills that were not activated yet (abs. count of 0) will not be printed in the chat.

Screenshot_4

One alternative to posting the info in the chat is by pressing Ctrl and Left Clicking on the player's stats in the party statistics window.

### CODE REVIEW

File: GWToolboxdll/Windows/PartyStatistics.cpp [hunk 1]
```diff
@@ -346,11 +335,19 @@ void PartyStatisticsWindow::Initialize() {
 
     ToolboxWindow::Initialize();
 
+    /* Reset party data on player add */
+    GW::StoC::RegisterPacketCallback<GW::Packet::StoC::PartyPlayerAdd>(&PartyPlayerAdd_Callback,
```
Comment (3vcloud) [inline GWToolboxdll/Windows/PartyStatistics.cpp]: Instead of having lambda function, declare `PartyClearCallback` as static with the relevent arguments.

File: GWToolboxdll/Windows/PartyStatistics.cpp [hunk 2]
```diff
@@ -24,11 +24,11 @@
 void PartyStatisticsWindow::GetSkillName(const uint32_t skill_id, char* const skill_name) {
     const auto& skill_data = GW::SkillbarMgr::GetSkillConstantData(skill_id);
 
-    const auto length = size_t{32};
-    wchar_t buffer[length] = {L'\0'};
+    const auto buffer_length = size_t{32};
```
Comment (3vcloud) [inline GWToolboxdll/Windows/PartyStatistics.cpp]: Don't use autos unless its actually useful, just declare it as `const size_t buffer_length = 32`

File: GWToolboxdll/Modules/ChatCommands.cpp [hunk 3]
```diff
@@ -421,7 +421,7 @@ void ChatCommands::Initialize() {
     });
     GW::Chat::CreateCommand(L"hero", ChatCommands::CmdHeroBehaviour);
     GW::Chat::CreateCommand(L"morale", ChatCommands::CmdMorale);
-    GW::Chat::CreateCommand(L"morale", ChatCommands::CmdSkillStatistics);
+    GW::Chat::CreateCommand(L"skillstats", ChatCommands::CmdSkillStatistics);
```
Comment (3vcloud) [inline GWToolboxdll/Modules/ChatCommands.cpp:424]: Put this into your module, not in `ChatCommands` - modules can be disabled so this would be a problem

File: GWToolboxdll/Windows/PartyStatistics.cpp [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+    if (nullptr == buffer) return;
+
+    const auto length = size_t{32};
+    GW::UI::AsyncDecodeStr(buffer, agent_name, length);
+}
+
+void PartyStatisticsWindow::ClearOnPartySizeChange() {
+    if (GW::Constants::InstanceType::Outpost != GW::Map::GetInstanceType()) return;
+
+    static auto last_party_size = static_cast<size_t>(-1);
```
Comment (3vcloud) [inline GWToolboxdll/Windows/PartyStatistics.cpp]: no auto

File: GWToolboxdll/Windows/PartyStatistics.cpp [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+    const auto party_member_found = agent_it != Instance().party_stats.end();
+
+    if (!party_member_found) return;
+
+    const auto agent_idx = std::distance(Instance().party_stats.begin(), agent_it);
+
+    auto& member_stats = Instance().party_stats[agent_idx];
+    auto& member_skill_counts = member_stats.skill_counts;
+
+    const auto skill_it = std::find_if(member_skill_counts.begin(), member_skill_counts.end(),
```
Comment (3vcloud) [inline GWToolboxdll/Windows/PartyStatistics.cpp]: No idea what any of the var types are looking at this code without an IDE to hold my hand - try not to use auto in this way too much

File: GWToolboxdll/Windows/PartyStatistics.cpp [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+            ImGui::Text("Hero data cannot be loded in outpost n/a");
+        }
+        return;
+    }
+
+    GetPlayerName(agent, agent_name);
+
+    snprintf(header_label, buffer_length, "%s", agent_name);
+
+    static auto is_open = true;
```
Comment (3vcloud) [inline GWToolboxdll/Windows/PartyStatistics.cpp]: no auto

File: GWToolboxdll/Windows/PartyStatistics.cpp [hunk 7]
```diff
@@ -25,19 +25,28 @@
...
 
-    wchar_t buffer[BUFFER_LENGTH] = {L'\0'};
+    wchar_t encoding_buffer[BUFFER_LENGTH] = {L'\0'};
+    wchar_t decoding_buffer[BUFFER_LENGTH] = {L'\0'};
 
     const auto encoding_length = size_t{16};
-    if (GW::UI::UInt32ToEncStr(skill_data.name, buffer, encoding_length)) {
-        GW::UI::AsyncDecodeStr(buffer, skill_name, BUFFER_LENGTH);
+    if (GW::UI::UInt32ToEncStr(skill_data.name, encoding_buffer, encoding_length)) {
+        GW::UI::AsyncDecodeStr(encoding_buffer, decoding_buffer, BUFFER_LENGTH);
```
Comment (3vcloud) [inline GWToolboxdll/Windows/PartyStatistics.cpp:36]: Don't call this synchronously and expect it to work all the time. Use `GuiUtils::EncStr` which wraps this

File: GWToolboxdll/Modules/ChatCommands.cpp [hunk 8]
```diff
@@ -349,30 +349,30 @@ void ChatCommands::Initialize() {
     const DWORD def_scale = 0x64000000;
     // Available Transmo NPCs
     // @Enhancement: Ability to target an NPC in-game and add it to this list via a GUI
-    npc_transmos = { 
-        {"charr", {163, def_scale, 0x0004c409, 0, 98820}}, 
-        {"reindeer", {5, def_scale, 277573, 277576, 32780}}, 
-        {"gwenpre", {244, def_scale, 116377, 116759, 98820}}, 
-        {"gwenchan", {245, def_scale, 116377, 283392, 98820}}, 
+    npc_transmos = {
```
Comment (3vcloud) [inline GWToolboxdll/Modules/ChatCommands.cpp:353]: Don't bother reformatting files that you haven't touched - its unnecessary chaff to check through without any functional change being made

**Review by 3vcloud — COMMENTED** [review]
TLDR; Don't use autos when it makes variables harder to identify the type, use static methods for some packet callbacks, don't clang format files that haven't changed functionally

File: GWToolboxdll/Windows/PartyStatistics.cpp [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+        if (static_cast<uint32_t>(0U) == total_num_skills) total_num_skills = 1U;
+
+        for (const auto [skill_id, count] : party_member_stats.skill_counts) {
+            char skill_name[buffer_length] = {'\0'};
+            GetSkillName(skill_id, skill_name);
+
+            const auto percentage = (static_cast<float>(count) / static_cast<float>(total_num_skills)) * 100.0F;
+
+            if (show_perc_values && show_abs_values) {
+                ImGui::Text("%s: %u\t-\t%-3.2f%%", skill_name, count, percentage);
```
Comment (3vcloud) [inline GWToolboxdll/Windows/PartyStatistics.cpp]: Align this in columns

File: GWToolboxdll/Windows/PartyStatistics.cpp (hunk 1 above)
Comment (franneck94) [reply GWToolboxdll/Windows/PartyStatistics.cpp]: Is there any special reason? Throughout the code base, I saw both.

File: GWToolboxdll/Modules/ChatCommands.cpp (hunk 3 above)
Comment (franneck94) [reply GWToolboxdll/Modules/ChatCommands.cpp:424]: What exactly do you mean? My command function is kind of the same strucutre than for example the CmdDamage function.

File: GWToolboxdll/Windows/PartyStatistics.cpp (hunk 7 above)
Comment (franneck94) [reply GWToolboxdll/Windows/PartyStatistics.cpp:36]: Should this then also be updated for the following modules?
Or was it inteded there?

Screenshot_6

File: GWToolboxdll/Windows/PartyStatistics.cpp (hunk 6 above)
Comment (franneck94) [reply GWToolboxdll/Windows/PartyStatistics.cpp]: Done.

File: GWToolboxdll/Windows/PartyStatistics.cpp (hunk 5 above)
Comment (franneck94) [reply GWToolboxdll/Windows/PartyStatistics.cpp]: Done.

File: GWToolboxdll/Windows/PartyStatistics.cpp (hunk 4 above)
Comment (franneck94) [reply GWToolboxdll/Windows/PartyStatistics.cpp]: Done.

File: GWToolboxdll/Windows/PartyStatistics.cpp (hunk 2 above)
Comment (franneck94) [reply GWToolboxdll/Windows/PartyStatistics.cpp]: Done.

File: GWToolboxdll/Windows/PartyStatistics.cpp (hunk 9 above)
Comment (3vcloud) [reply GWToolboxdll/Windows/PartyStatistics.cpp]: If it doesn't break anything then a PR that has the dependency pulled up to date would be welcome, but I doubt it'll be that simple ;)

File: GWToolboxdll/Windows/PartyStatistics.cpp (hunk 7 above)
Comment (3vcloud) [reply GWToolboxdll/Windows/PartyStatistics.cpp:36]: There are some edge cases but yes, these lines will be refactored when I get around to it. The class I made makes it a bit more manageable, but you'll need to figure out how to keep a reference to these decoded strings across frames

File: GWToolboxdll/Modules/ChatCommands.cpp (hunk 3 above)
Comment (3vcloud) [reply GWToolboxdll/Modules/ChatCommands.cpp:424]: If you were to disable the damage module in tb settings, restart toolbox and then run the damage command, it wouldnt work because the module isn't running. Its compiled in with toolbox so you won't crash or anything but keeping module functions as separate as you can means that if you add an optional module from another DLL at runtime its all self contained

File: GWToolboxdll/Windows/PartyStatistics.cpp (hunk 1 above)
Comment (3vcloud) [reply GWToolboxdll/Windows/PartyStatistics.cpp]: If you're not doing any logic in the lambda its pointless - depending on the compiler, you've got 2 jumps instead of 1 and the code would be more readable. Your generic packet handler callback is a good example of when its OK to use it because you're passing in different vars depending on the packet

File: GWToolboxdll/Windows/PartyStatistics.cpp (hunk 9 above)
Comment (franneck94) [reply GWToolboxdll/Windows/PartyStatistics.cpp]: For now i use the old Columns API.

Is this layout okay for you?

Screenshot_7

File: GWToolboxdll/Modules/ChatCommands.cpp (hunk 8 above)
Comment (franneck94) [reply GWToolboxdll/Modules/ChatCommands.cpp:353]: Noted for the future. Should i revert it for now?

File: GWToolboxdll/Windows/PartyStatistics.cpp (hunk 9 above)
Comment (3vcloud) [reply GWToolboxdll/Windows/PartyStatistics.cpp]: Looks good yeah, I'd keep last 2 columns fixed with and the first can be whatever is left. Don't forget to scale with global font scale

File: GWToolboxdll/Modules/ChatCommands.cpp (hunk 8 above)
Comment (3vcloud) [reply GWToolboxdll/Modules/ChatCommands.cpp:353]: Yep revert all of these files for now, makes it easier to track feature changes

File: GWToolboxdll/Windows/PartyStatistics.cpp (hunk 9 above)
Comment (franneck94) [reply GWToolboxdll/Windows/PartyStatistics.cpp]: So it isnt enough to just use the current window width?

```cpp
const float width = ImGui::GetWindowWidth();
ImGui::Columns(3, table_name, false);
ImGui::SetColumnWidth(0, width * 0.6F);
ImGui::SetColumnWidth(1, width * 0.2F);
ImGui::SetColumnWidth(2, width * 0.2F);
```

Comment (3vcloud) [reply GWToolboxdll/Windows/PartyStatistics.cpp]: Guess so, I usually keep columns that won't contain string text to a fixed width I.e. when you expand the window, only the skill name fills the extra space but this works fine

File: GWToolboxdll/Modules/ChatCommands.cpp (hunk 3 above)
Comment (franneck94) [reply GWToolboxdll/Modules/ChatCommands.cpp:424]: I still dont get your point. Are you refering to exactly this line in the review, or to the CmdSkillStatistics function itself?

Comment (3vcloud) [reply GWToolboxdll/Modules/ChatCommands.cpp:424]: This line needs to be in your module, not the chat commands module

Comment (franneck94) [reply GWToolboxdll/Modules/ChatCommands.cpp:424]: Okay i will do this.

But sorry for being stupid, but why is the GW::Chat::CreateCommand(L"dmg", ChatCommands::CmdDamage); command then in the ChatCommands.cpp and not in the PartyDamage.cpp?
Or is this just an outdated code style?

Comment (3vcloud) [reply GWToolboxdll/Modules/ChatCommands.cpp:424]: Yep, there's an ongoing task to allow modules to be side loaded into toolbox, which means the functionality of each module would be self contained. This includes registering hooks for chat commands

File: GWToolboxdll/Windows/PartyStatistics.cpp [hunk 10]
```diff
@@ -25,28 +25,32 @@
...
+        skill_name_enc = (*it).second.encoder;
     }
 
-    std::wstring skill_name{decoding_buffer};
+    if (!skill_name_enc || skill_name_enc->IsDecoding()) return UNKNOWN_SKILL_NAME;
 
-    if (NONE_SKILL_NAME == skill_name) {
-        skill_name = std::wstring{UNKNOWN_SKILL_NAME};
-    }
+    const std::wstring result = skill_name_enc->wstring();
```
Comment (3vcloud) [inline GWToolboxdll/Windows/PartyStatistics.cpp:45]: I think this would do a copy

File: GWToolboxdll/Windows/PartyStatistics.cpp [hunk 11]
```diff
@@ -25,28 +25,32 @@
...
 
-    return skill_name;
+    if (result.empty()) return UNKNOWN_SKILL_NAME;
+
+    delete skill_name_enc;
```
Comment (3vcloud) [inline GWToolboxdll/Windows/PartyStatistics.cpp:49]: You'll leave a hanging pointer if skill name hasn't decoded yet

Comment (franneck94) [reply GWToolboxdll/Windows/PartyStatistics.cpp:49]: While a skill name is still decoding no new heap allocation will happen for this skill und the already created pointer will be used (logic is in the if/else at the beginning of the function).

File: GWToolboxdll/Windows/PartyStatistics.cpp (hunk 10 above)
Comment (franneck94) [reply GWToolboxdll/Windows/PartyStatistics.cpp:45]: Should i do a move after decoding is finished? A Reference wouldnt work since i will delete the pointer at the end of the function.

Comment (franneck94) [reply GWToolboxdll/Windows/PartyStatistics.cpp:45]: Should i do a move after decoding is finished? A Reference wouldnt work since i will delete the pointer at the end of the function.

Comment (franneck94) [reply GWToolboxdll/Windows/PartyStatistics.cpp:45]: Do you just mean i should only do a copy after decoding is finished?

File: GWToolboxdll/Windows/PartyStatistics.cpp (hunk 11 above)
Comment (3vcloud) [reply GWToolboxdll/Windows/PartyStatistics.cpp:49]: Did a quick amendment @ https://github.com/3vcloud/GWToolboxpp/commit/e3849bacf65cec2871be4df3c16dac21ae25f2e7 to make string decoding a little less confusing
